File: java/leetcode/src/Task101.java
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class Task101 {

    public boolean isSymmetric(TreeNode root) {
        if(root == null){
            return true;
        }
        //condition

        return  isMirror(root.left,root.right);
    }

    boolean isMirror(TreeNode t1,TreeNode t2){
        if(t1 == null){
            return t2 == null;
        }
        if(t2 == null){
            return t1 == null;
        }

        return t1.val == t2.val
                && isMirror(t1.left,t2.right)
                && isMirror(t1.right,t2.left);

    }


    public boolean isSymmetric2(TreeNode root) {
        if(root == null){
            return true;
        }
        //condition
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root.left);
        queue.offer(root.right);

        while (!queue.isEmpty()){
            TreeNode t1 = queue.poll();
            TreeNode t2 = queue.poll();
            if (t1 == null && t2 == null) continue;
            if (t1 == null || t2 == null) return false;
            if(t1.val != t2.val){
                return false;
            }
            queue.addAll(Arrays.asList(t1.left,t2.right,t1.right,t2.left));

        }

        return true;
    }


    @Test
    void test(){
        TreeNode root = new TreeNode(1);
        root.left(2).right(2);
        root.left.left(3).right(4);
        root.right.left(4).right(3);

        System.out.println(isSymmetric2(root));
    }
}

File: py/Task101.py
# Definition for a binary tree node.
from queue import Queue


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    # dfs
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True

        def isSymmetric(l: TreeNode, r: TreeNode) -> bool:
            if not l:
                return not r
            if not r:
                return False
            if l.val != r.val:
                return False

            return isSymmetric(l.left, r.right) and isSymmetric(l.right, r.left)

        return isSymmetric(root.left, root.right)

    # iter
    def isSymmetric2(self, root: TreeNode) -> bool:
        if not root:
            return True
        q = Queue()
        q.put(root.left)
        q.put(root.right)
        while not q.empty():
            l = q.get()
            r = q.get()
            if not l and not r:
                continue
            if not l or not r or l.val != r.val:
                return False
            q.put(l.right)
            q.put(r.left)
            q.put(l.left)
            q.put(r.right)
        return True


obj = Solution()
root = TreeNode(1)
root.left, root.right = TreeNode(2), TreeNode(2)
root.left.left, root.left.right = TreeNode(3), TreeNode(4)
root.right.left, root.right.right = TreeNode(4), TreeNode(3)
print(obj.isSymmetric2(root))
